File: leetcode/python/8.Atoi.py
class ATOI:
    @staticmethod
    def atoi(s):
        st = s.strip()

        if (st is None or len(st) == 0):
            return 0

        res = 0
        isNegative = False
        startIndex = 0

        if st[0] == '+' or st[0] == '-':
            startIndex += 1
        if st[0] == '-':
            isNegative = True

        print(startIndex, isNegative)

        i = startIndex

        while i < len(st):
            if st[i] < '0' or st[i] > '9':
                break
            digit = (int)(st[i])
            res = res * 10 + digit
            i += 1

        if isNegative:
            res = -res

        if res > 2 ^ 31 - 1:
            return 2 ^ 31

        if res < -2 ^ 31 + 1:
            return -2 ^ 31

        return res


# import sys
# sys.maxsize

print(ATOI.atoi("-23"))

File: leetcode/c++/8.Atoi.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <bits/stdc++.h>

using namespace std;

class ATOI{
public:

  static string trim(const string &s){
  auto start = s.begin();
  while(start != s.end() && isspace(*start)){
     start++;
  }
  auto end = s.end();
  do{
    end--;
  }while(distance(start,end)>0 && isspace(*end));
  return string(start,end+1);
  }

  static int atoi(string s){
        string str = trim(s);
        double result = 0;
        bool isNegative = false;
        int startIndex = 0;
        if (str[0] == '+' || str[0] == '-') {
            startIndex++;
        }
        if (str[0] == '-') {
            isNegative = true;
        }
        for (int i = startIndex; i < str.length(); i++) {
            if (str[i] < '0' || str[i] > '9') {
                break;
            }
            int digitValue = (int) (str[i] - '0');
            result = result * 10 + digitValue;
        }

        if (isNegative) {
            result = -result;
        }
        if (result > INT_MAX) {
            return INT_MAX;
        }
        if (result < INT_MIN) {
            return INT_MIN;
        }

        return (int) result;
  }

};

int main() {
  cout << ATOI::atoi(" 42 Hello World! ")<<endl;
  return 0;
}